src/mococrw/dilithium.h

```diff
@@ -0,0 +1,392 @@
+
+private:
+    std::shared_ptr<DilithiumKeyImpl> _key;
+};
+
+/**
+ * @brief Dilithium representation of AsymmetricPublicKey
+ *
+ * @see AsymmetricPublicKey
+ *
+ * This shall be replaced by AsymmetricPublicKey once Dilithium is supported by the openssl version
+ * used by MoCOCrW.
+ * The interface is similar to the one of AsymmetricPublicKey.
+ *
+ * For documentation see AsymmetricPublicKey
+ */
+class DilithiumAsymmetricPublicKey
+{
+public:
+    DilithiumAsymmetricPublicKey(std::shared_ptr<DilithiumKeyImpl> key) : _key{std::move(key)} {}
+
+    /**
+     * @brief Parses an ASN.1 struct in DER format at returns a public key object
+     *
+     * @param asn1Data The ASN.1 struct in DER format
+     * @return the DilithiumAsymmetricPublicKey object created form the DER data
+     * @throws MoCOCrWException if the struct cannot be parsed or if the key is invalid
+     */
+    static DilithiumAsymmetricPublicKey readPublicKeyfromDER(const std::vector<uint8_t> &asn1Data);
+
+    /**
+     * Getters for the internal dilithium implementation objet.
+     */
+    inline DilithiumKeyImpl *internal() { return _key._internal().get(); }
+    inline const DilithiumKeyImpl *_internal() const { return _key._internal().get(); }
+
+    /**
+     * Gets the type of the asymmetric key or key pair, @see AsymmetricKey::KeyTypes for the
+     * supported types.
+     * @return the type of the asymmetric key.
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error
+     */
+    AsymmetricKey::KeyTypes getType() const { return _key.getType(); }
+
+    /**
+     * Gets the specification of the asymmetric key in usage
+     *
+     * @return the specification of the key in use.
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error
+     */
+    std::unique_ptr<DilithiumAsymmetricKey::Spec> getKeySpec() const { return _key.getKeySpec(); }
+
+    /**
+     * Gets the size of the Asymmetric key in bits
+     * @return the size of the key in bits
+     */
+    int getKeySize() const { return _key.getKeySize(); }
+
+    inline bool operator==(const DilithiumAsymmetricPublicKey &rhs) const
+    {
+        return *_internal() == *rhs._internal();
+    }
+    inline bool operator!=(const DilithiumAsymmetricPublicKey &rhs) const
+    {
+        return !(*this == rhs);
+    }
+
+protected:
+    DilithiumAsymmetricPublicKey(DilithiumAsymmetricKey &&key) : _key{std::move(key)} {}
+
+    DilithiumAsymmetricKey _key;
+};
+
+/**
+ * @brief Dilithium representation of AsymmetricKeypair
+ *
+ * @see AsymmetricKeypair
+ *
+ * This shall be replaced by AsymmetricKeypair once Dilithium is supported by the openssl version
+ * used by MoCOCrW.
+ * The interface is similar to the one of AsymmetricKeypair.
+ *
+ * For documentation see AsymmetricKeypair
+ */
+class DilithiumAsymmetricKeypair : public DilithiumAsymmetricPublicKey
+{
+public:
+    DilithiumAsymmetricKeypair(std::shared_ptr<DilithiumKeyImpl> keypair)
```

Comment:
Probably we can make this constructor private. This shouldn't be used. Only the reader or generate function should be used to create new key objects. Less potential for future incompatibilities…

Reply:
This requires two friend classes. Did it as extra commit. If you like it, I will fixup the commit, if not, it can be easily removed.